src/Libraries/Microsoft.Extensions.DataIngestion/Writers/VectorStoreWriter.cs

```diff
@@ -0,0 +1,201 @@
+
+    /// <inheritdoc/>
+    public override async Task WriteAsync(IAsyncEnumerable<IngestionChunk<T>> chunks, CancellationToken cancellationToken = default)
+    {
+        _ = Throw.IfNull(chunks);
+
+        cancellationToken.ThrowIfCancellationRequested();
+
+        IReadOnlyList<object>? preExistingKeys = null;
+        await foreach (IngestionChunk<T> chunk in chunks.WithCancellation(cancellationToken))
+        {
+            if (_vectorStoreCollection is null)
+            {
+                _vectorStoreCollection = _vectorStore.GetDynamicCollection(_options.CollectionName, GetVectorStoreRecordDefinition(chunk));
+
+                await _vectorStoreCollection.EnsureCollectionExistsAsync(cancellationToken).ConfigureAwait(false);
+            }
+
+            // We obtain the IDs of the pre-existing chunks for given document,
+            // and delete them after we finish inserting the new chunks.
+            // To avoid a situation where we delete the chunks and then fail to insert the new ones.
+            preExistingKeys ??= await GetPreExistingChunksIdsAsync(chunk.Document, cancellationToken).ConfigureAwait(false);
+
+            var key = Guid.NewGuid();
+            Dictionary<string, object?> record = new()
+            {
+                [KeyName] = _keysAreStrings ? key.ToString() : key,
+                [ContentName] = chunk.Content,
+                [EmbeddingName] = chunk.Content,
+                [ContextName] = chunk.Context,
+                [DocumentIdName] = chunk.Document.Identifier,
+            };
+
+            if (chunk.HasMetadata)
+            {
+                foreach (var metadata in chunk.Metadata)
+                {
+                    record[metadata.Key] = metadata.Value;
+                }
+            }
+
+            await _vectorStoreCollection.UpsertAsync(record, cancellationToken).ConfigureAwait(false);
+        }
+
+        if (preExistingKeys?.Count > 0)
+        {
+            await _vectorStoreCollection!.DeleteAsync(preExistingKeys, cancellationToken).ConfigureAwait(false);
+        }
+    }
+
+    /// <inheritdoc/>
+    protected override void Dispose(bool disposing)
+    {
+        try
+        {
+            _vectorStore.Dispose();
+            _vectorStoreCollection?.Dispose();
+        }
+        finally
+        {
+            base.Dispose(disposing);
+        }
+    }
+
+    private VectorStoreCollectionDefinition GetVectorStoreRecordDefinition(IngestionChunk<T> representativeChunk)
+    {
+        VectorStoreCollectionDefinition definition = new()
+        {
+            Properties =
+            {
+                new VectorStoreKeyProperty(KeyName, _keysAreStrings ? typeof(string) : typeof(Guid)),
+
+                // By using T as the type here we allow the vector store
+                // to handle the conversion from T to the actual vector type it supports.
+                new VectorStoreVectorProperty(EmbeddingName, typeof(T), _dimensionCount)
```

Comment:
Just checking: T corresponds to IngestionChunk.Content, right? If so, then the user is responsible for passing in a VectorStore that has been configured with the appropriate EmbeddingGenerator (that supports converting whatever T is being used here to e.g. `Embedding<float>`), right?